buzzmobile/tests/test_utils/mock_nodes.py

```diff
@@ -4,114 +4,112 @@
 """
 
 import contextlib
+import cPickle as pickle
+import subprocess
 import time
-import rospy
+from StringIO import StringIO
+
+import rosnode
+
 
 class TimeoutError(Exception):
     """Py3 shim, represents no response from a syscall.
     """
     pass
 
 
-
 class NoMessage(Exception):
     """Exception for a lack of message in a Node.
     """
     pass
 
-class MockNode(object):
+class MockPublisher(object):
     """Mock of a node object for testing.
     """
-    def __init__(self, topic, msg_type, node):
+    def __init__(self, topic, msg_type, queue_size):
         self.topic = topic
         self.msg_type = msg_type
-        self.node = node
+        pub_data = pickle.dumps((topic, msg_type, queue_size))
+        location = './tests/test_utils/publisher.py'
+        self.proc = subprocess.Popen([location, pub_data],
+                stdin=subprocess.PIPE)
 
     def send(self, value):
         """Sends data to be publoshed to the mocked topic.
         """
-        self.node.publish(value)
+        value.serialize(self.proc.stdin)
         # Just long enough to prevent out of order
-        time.sleep(.1)
+        time.sleep(.2)
+
+    def kill(self):
+        """Kills the publisher ros node process.
+        """
+        self.proc.kill()
+
 
 @contextlib.contextmanager
 def mock_pub(topic, rosmsg_type, queue_size=1):
     """Mocks a node and cleans it up when done.
     """
-    pub = rospy.Publisher(topic, rosmsg_type, queue_size=queue_size)
-    yield MockNode(topic, rosmsg_type, pub)
-    pub.unregister()
+    pub = MockPublisher(topic, rosmsg_type, queue_size)
+    no_ns = topic.split('/')[-1]
+    while not any(nn.split('/')[-1].startswith(
+        ''.join(['mock_publish_', no_ns])) for nn in rosnode.get_node_names()):
+        time.sleep(.1)
+    try:
+        yield pub
+    finally:
+        pub.kill()
 
 
-class TestNode(object):
+class MockListener(object):
     """Wrapper around a node used for testing.
     """
 
     def __init__(self, topic, msg_type):
         self.topic = topic
         self.msg_type = msg_type
-        self.received = False
+        # get port and rosmaster uri cleanly
+
+        # do this better, somehow!
+        location = './tests/test_utils/listener.py'
+        self.proc = subprocess.Popen([location,
```

Comment:
Generally though, because this is such a weird system, can you make a `buzzmobile/tests/test_utils/readme.md` explaining how it works?

"design doc pls mr intern"